File: app/services/mercadopago.py
"""
Mercado Pago OAuth + lectura de movimientos.
Requiere: MP_CLIENT_ID, MP_CLIENT_SECRET
"""
import httpx
from datetime import datetime, date, timedelta
from typing import List, Dict
from ..config import get_settings

def get_oauth_url(state: str) -> str:
    settings = get_settings()
    redirect = f"{settings.app_url}/api/auth/mp/callback"
    return (
        "https://auth.mercadopago.com/authorization"
        f"?client_id={settings.mp_client_id}"
        f"&redirect_uri={redirect}"
        f"&response_type=code"
        f"&state={state}"
    )

async def exchange_code(code: str) -> dict:
    import httpx
    settings = get_settings()
    async with httpx.AsyncClient() as client:
        r = await client.post("https://api.mercadopago.com/oauth/token", data={
            "grant_type": "authorization_code",
            "client_id": settings.mp_client_id,
            "client_secret": settings.mp_client_secret,
            "code": code,
            "redirect_uri": f"{settings.app_url}/api/auth/mp/callback",
        })
        r.raise_for_status()
        return r.json()

async def fetch_movements(access_token: str, days_back: int = 30) -> List[Dict]:
    """Lee movimientos de Mercado Pago vía /v1/payments/search.

    Nota: /v1/collections/search fue dado de baja por MP (devuelve 404).
    payments/search trae pagos donde el usuario participa; la dirección se
    determina comparando collector_id / payer.id con el id del usuario.
    """
    headers = {"Authorization": f"Bearer {access_token}"}
    since = (datetime.utcnow() - timedelta(days=days_back)).strftime("%Y-%m-%dT00:00:00.000-00:00")
    until = (datetime.utcnow() + timedelta(days=1)).strftime("%Y-%m-%dT00:00:00.000-00:00")
    results = []

    async with httpx.AsyncClient(timeout=30) as client:
        me_resp = await client.get("https://api.mercadopago.com/users/me", headers=headers)
        if me_resp.status_code != 200:
            raise Exception(f"MP users/me error {me_resp.status_code}: {me_resp.text}")
        my_id = str(me_resp.json().get("id", ""))

        url_payments = "https://api.mercadopago.com/v1/payments/search"
        offset = 0
        while True:
            resp = await client.get(url_payments, headers=headers, params={
                "sort": "date_approved", "criteria": "desc",
                "range": "date_approved", "begin_date": since, "end_date": until,
                "limit": 50, "offset": offset
            })
            if resp.status_code not in (200, 201):
                raise Exception(f"MP payments API error {resp.status_code}: {resp.text}")
            data = resp.json()
            items = data.get("results", [])
            if not items:
                break
            for item in items:
                if item.get("status") != "approved":
                    continue
                raw_date = item.get("date_approved", "")[:10]
                if not raw_date:
                    continue

                collector_id = str(item.get("collector_id")
                                   or (item.get("collector") or {}).get("id") or "")
                payer = item.get("payer") or {}
                is_income = bool(my_id) and collector_id == my_id

                description = (item.get("description") or "").strip()
                is_transfer = item.get("payment_type_id") == "money_transfer"

                if is_income:
                    merchant = description or payer.get("email", "") or "Transferencia recibida"
                    needs_review = is_transfer and not description
                else:
                    merchant = description or item.get("payment_method_id", "")
                    needs_review = False

                results.append({
                    "id": f"mp_{item['id']}",
                    "source": "mercadopago",
                    "tx_type": "income" if is_income else "expense",
                    "amount": float(item.get("transaction_amount", 0)),
                    "currency": item.get("currency_id", "ARS"),
                    "date": raw_date,
                    "month": raw_date[:7],
                    "merchant": merchant,
                    "provider": "MercadoPago",
                    "needs_review": needs_review,
                    "raw_reference": str(item.get("id")),
                })
            total = data.get("paging", {}).get("total", 0)
            offset += 50
            if offset >= total:
                break

    return results

async def refresh_token(token: str) -> dict:
    import httpx
    settings = get_settings()
    async with httpx.AsyncClient() as client:
        r = await client.post("https://api.mercadopago.com/oauth/token", data={
            "grant_type": "refresh_token",
            "client_secret": settings.mp_client_secret,
            "refresh_token": token,
        })
        r.raise_for_status()
        return r.json()

File: packages/shared/mercadopago.js
/** Mercado Pago OAuth + lectura de movimientos. */
import { settings } from "./config.js";
import { fromMp } from "./paymentMethod.js";
export function getOauthUrl(state) {
  const params = new URLSearchParams({
    client_id: settings.mpClientId,
    redirect_uri: `${settings.appUrl}/api/auth/mp/callback`,
    response_type: "code",
    platform_id: "mp",
    state
  });
  return `https://auth.mercadopago.com/authorization?${params}`;
}
export async function exchangeCode(code) {
  const r = await fetch("https://api.mercadopago.com/oauth/token", {
    method: "POST",
    headers: {
      "content-type": "application/x-www-form-urlencoded"
    },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      client_id: settings.mpClientId,
      client_secret: settings.mpClientSecret,
      code,
      redirect_uri: `${settings.appUrl}/api/auth/mp/callback`
    })
  });
  if (!r.ok) throw new Error(`MP token error ${r.status}`);
  return r.json();
}
export async function fetchMovements(accessToken, daysBack = 180) {
  const headers = {
    Authorization: `Bearer ${accessToken}`
  };
  const since = new Date(Date.now() - daysBack * 86400000).toISOString().slice(0, 10) + "T00:00:00.000-00:00";
  const until = new Date(Date.now() + 86400000).toISOString().slice(0, 10) + "T00:00:00.000-00:00";
  const results = [];
  const meResp = await fetch("https://api.mercadopago.com/users/me", {
    headers
  });
  if (!meResp.ok) throw new Error(`MP users/me error ${meResp.status}: ${await meResp.text()}`);
  const me = await meResp.json();
  const myId = String(me.id || "");
  let offset = 0;
  while (true) {
    const url = new URL("https://api.mercadopago.com/v1/payments/search");
    url.search = new URLSearchParams({
      sort: "date_approved",
      criteria: "desc",
      range: "date_approved",
      begin_date: since,
      end_date: until,
      limit: "50",
      offset: String(offset)
    }).toString();
    const resp = await fetch(url, {
      headers
    });
    if (![200, 201].includes(resp.status)) throw new Error(`MP payments API error ${resp.status}: ${await resp.text()}`);
    const data = await resp.json();
    const items = data.results || [];
    if (!items.length) break;
    for (const item of items) {
      if (item.status !== "approved") continue;
      const rawDate = String(item.date_approved || "").slice(0, 10);
      if (!rawDate) continue;
      const collectorId = String(item.collector_id ?? item.collector?.id ?? "");
      const payer = item.payer || {};
      const isIncome = Boolean(myId) && collectorId === myId;
      const description = String(item.description || "").trim();
      const paymentTypeId = item.payment_type_id;
      const isTransfer = paymentTypeId === "money_transfer";
      const paymentMethod = fromMp(paymentTypeId, item.payment_method_id);
      let merchant;
      let needsReview;
      if (isIncome) {
        merchant = description || payer.email || "Transferencia recibida";
        needsReview = isTransfer && !description;
      } else {
        merchant = description || item.payment_method_id || "";
        needsReview = false;
      }
      results.push({
        id: `mp_${item.id}`,
        source: "mercadopago",
        tx_type: isIncome ? "income" : "expense",
        amount: Number(item.transaction_amount || 0),
        currency: item.currency_id || "ARS",
        date: rawDate,
        month: rawDate.slice(0, 7),
        merchant,
        provider: "MercadoPago",
        payment_method: paymentMethod,
        needs_review: needsReview,
        raw_reference: String(item.id)
      });
    }
    const total = data.paging?.total || 0;
    offset += 50;
    if (offset >= total) break;
  }
  return results;
}
export async function refreshToken(token) {
  const r = await fetch("https://api.mercadopago.com/oauth/token", {
    method: "POST",
    headers: {
      "content-type": "application/x-www-form-urlencoded"
    },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      client_secret: settings.mpClientSecret,
      refresh_token: token
    })
  });
  if (!r.ok) throw new Error(`MP refresh error ${r.status}`);
  return r.json();
}
